letsencrypt/renewer.py

```diff
@@ -0,0 +1,132 @@
+
+    # TODO: handle partial success (some names can be renewed but not
+    #       others)
+    # TODO: handle obligatory key rotation vs. optional key rotation vs.
+    #       requested key rotation
+    if "renewalparams" not in cert.configfile:
+        # TODO: notify user?
+        return False
+    renewalparams = cert.configfile["renewalparams"]
+    if "authenticator" not in renewalparams:
+        # TODO: notify user?
+        return False
+    # Instantiate the appropriate authenticator
+    plugins = plugins_disco.PluginsRegistry.find_all()
+    config = configuration.NamespaceConfig(AttrDict(renewalparams))
+    # XXX: this loses type data (for example, the fact that key_size
+    #      was an int, not a str)
+    config.rsa_key_size = int(config.rsa_key_size)
+    try:
+        authenticator = plugins[renewalparams["authenticator"]]
+    except KeyError:
+        # TODO: Notify user? (authenticator could not be found)
+        return False
+    authenticator = authenticator.init(config)
+
+    authenticator.prepare()
+    account = client.determine_account(config)
+    # TODO: are there other ways to get the right account object, e.g.
+    #       based on the email parameter that might be present in
+    #       renewalparams?
+
+    our_client = client.Client(config, account, authenticator, None)
+    with open(cert.version("cert", old_version)) as f:
+        sans = crypto_util.get_sans_from_cert(f.read())
+    new_cert, new_key, new_chain = our_client.obtain_certificate(sans)
+    if new_cert and new_key and new_chain:
+        # XXX: Assumes that there was no key change.  We need logic
+        #      for figuring out whether there was or not.  Probably
+        #      best is to have obtain_certificate return None for
+        #      new_key if the old key is to be used (since save_successor
+        #      already understands this distinction!)
+        return cert.save_successor(old_version, new_cert, new_key, new_chain)
+        # TODO: Notify results
+    else:
+        # TODO: Notify negative results
+        return False
+    # TODO: Consider the case where the renewal was partially successful
+    #       (where fewer than all names were renewed)
+
+
+def main(config=constants.RENEWER_DEFAULTS):
+    """main function for autorenewer script."""
+    # TODO: Distinguish automated invocation from manual invocation,
+    #       perhaps by looking at sys.argv[0] and inhibiting automated
+    #       invocations if /etc/letsencrypt/renewal.conf defaults have
+    #       turned it off. (The boolean parameter should probably be
+    #       called renewer_enabled.)
+
+    # This attempts to read the renewer config file and augment or replace
+    # the renewer defaults with any options contained in that file.  If
+    # renewer_config_file is undefined or if the file is nonexistent or
+    # empty, this .merge() will have no effect.
+    config.merge(configobj.ConfigObj(config.get("renewer_config_file", "")))
```

Comment:
If someone calls `main()` (without parameters) this line will mutate `constants.RENEWER_DEFAULTS` which was meant to be constant in the first place